internal/octant/apply_yaml.go

```diff
@@ -0,0 +1,221 @@
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package octant
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/vmware-tanzu/octant/internal/cluster"
+	"github.com/vmware-tanzu/octant/pkg/action"
+	"github.com/vmware-tanzu/octant/pkg/log"
+	"github.com/vmware-tanzu/octant/pkg/store"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	sigyaml "sigs.k8s.io/yaml"
+)
+
+// ApplyYaml creates a yaml applier
+type ApplyYaml struct {
+	logger        log.Logger
+	objectStore   store.Store
+	clusterClient cluster.ClientInterface
+}
+
+var _ action.Dispatcher = (*ApplyYaml)(nil)
+
+// NewApplyYaml creates an instance of ApplyYaml
+func NewApplyYaml(logger log.Logger, objectStore store.Store, clusterClient cluster.ClientInterface) *ApplyYaml {
+	return &ApplyYaml{
+		logger:        logger,
+		objectStore:   objectStore,
+		clusterClient: clusterClient,
+	}
+}
+
+// ActionName returns the name of this action
+func (p *ApplyYaml) ActionName() string {
+	return ActionApplyYaml
+}
+
+// Handle applies the requested yaml to the cluster
+func (p *ApplyYaml) Handle(ctx context.Context, alerter action.Alerter, payload action.Payload) error {
+	p.logger.With("payload", payload).Debugf("received action payload")
+	request, err := applyYamlRequestFromPayload(payload)
+	if err != nil {
+		return errors.Wrap(err, "convert payload to apply yaml request")
+	}
+	p.logger.Debugf("%s", request)
+
+	results := []string{}
+	err = request.WithDoc(func(doc map[string]interface{}) error {
+		p.logger.Debugf("apply resource %#v", doc)
+
+		unstructuredObj := &unstructured.Unstructured{Object: doc}
+		key, err := store.KeyFromObject(unstructuredObj)
+		if err != nil {
+			return err
+		}
+		gvr, namespaced, err := p.clusterClient.Resource(key.GroupVersionKind().GroupKind())
+		if err != nil {
+			return fmt.Errorf("unable to discover resource: %w", err)
+		}
+		if namespaced && key.Namespace == "" {
+			unstructuredObj.SetNamespace(request.Namespace)
+			key.Namespace = request.Namespace
+		}
+
+		if _, err := p.objectStore.Get(ctx, key); err != nil {
+			if !kerrors.IsNotFound(err) {
+				// unexpected error
+				return fmt.Errorf("unable to get resource: %w", err)
+			}
+
+			// create object
+			err := p.objectStore.Create(ctx, &unstructured.Unstructured{Object: doc})
+			if err != nil {
+				return fmt.Errorf("unable to create resource: %w", err)
+			}
+
+			result := fmt.Sprintf("Created %s (%s) %s", key.Kind, key.APIVersion, key.Name)
+			if namespaced {
+				result = fmt.Sprintf("%s in %s", result, key.Namespace)
+			}
+			results = append(results, result)
+
+			return nil
+		}
+
+		// update object
+		unstructuredYaml, err := sigyaml.Marshal(doc)
+		if err != nil {
+			return fmt.Errorf("unable to marshal resource as yaml: %w", err)
+		}
+		True := true
```

Comment:
I understand why and even though we never actually change the value, I'd rather see this called `withForce` or even `Force` because then I at least get a hint at what this is doing. I'd like to see it moved to right before the if/else that is using it.